src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumMiniWallet.kt

```diff
@@ -0,0 +1,188 @@
+    private var scriptHashes: Map<ByteVector32, String> = emptyMap()
+
+    // the mailbox of this "actor"
+    private val mailbox: Channel<WalletCommand> = Channel(Channel.BUFFERED)
+
+    internal fun addAddress(bitcoinAddress: String, privateKey: PrivateKey? = null) {
+        launch {
+            mailbox.send(WalletCommand.Companion.AddAddress(bitcoinAddress, privateKey))
+        }
+    }
+
+    fun addPay2wpkhAddress(privateKey: PrivateKey) {
+        launch {
+            val bitcoinAddress = Bitcoin.computeP2WpkhAddress(privateKey.publicKey(), Block.TestnetGenesisBlock.hash)
+            mailbox.send(WalletCommand.Companion.AddAddress(bitcoinAddress, privateKey))
+        }
+    }
+
+    fun addWatchOnlyAddress(bitcoinAddress: String) {
+        addAddress(bitcoinAddress)
+    }
+
+    init {
+        launch {
+            // listen to connection events
+            client.connectionState
+                .filterIsInstance<Connection.ESTABLISHED>()
+                .collect { mailbox.send(WalletCommand.Companion.ElectrumConnected) }
+        }
+        launch {
+            // listen to subscriptions events
+            client.notifications
+                .filterIsInstance<ElectrumResponse>()
+                .collect { mailbox.send(WalletCommand.Companion.ElectrumNotification(it)) }
+        }
+
+        launch {
+            mailbox.consumeAsFlow().collect {
+                when (it) {
+                    is WalletCommand.Companion.ElectrumConnected -> {
+                        logger.info { "electrum connected" }
+                        scriptHashes.values.forEach { subscribe(it) }
+                    }
+                    is WalletCommand.Companion.ElectrumNotification -> {
+                        // NB: we ignore responses for unknown script_hashes (electrum client doesn't maintain a list of subscribers so we receive all subscriptions)
+                        when (val msg = it.msg) {
+                            is ScriptHashSubscriptionResponse -> {
+                                scriptHashes[msg.scriptHash]?.let { bitcoinAddress ->
+                                    if (msg.status.isNotEmpty()) {
+                                        logger.info { "non-empty status for address=$bitcoinAddress, requesting utxos" }
+                                        client.sendElectrumRequest(ScriptHashListUnspent(msg.scriptHash))
+                                    }
+                                }
+                            }
+                            is ScriptHashListUnspentResponse -> {
+                                scriptHashes[msg.scriptHash]?.let { address ->
+                                    val newUtxos = msg.unspents.minus((_walletStateFlow.value.addresses[address] ?: emptyList()).toSet())
+                                    // request new parent txs
+                                    newUtxos.forEach { utxo -> client.sendElectrumRequest(GetTransaction(utxo.tx_hash)) }
```

Comment:
How can that work? `GetTransaction` expects the `txid`, not the `tx_hash`?
Or is Electrum's `tx_hash` actually the `txid` (which would be very confusing...)?

Reply:
> Or is Electrum's tx_hash actually the txid (which would be very confusing...)?

Unfortunately yes...

Reply:
I think we should make them private indeed, and we should have tests to very explicitly show that this is the `txid`...

Reply:
WDYT? This is _the_ proper way:

```kotlin
data class UnspentItem(@SerialName("tx_hash") val txid: ByteVector32, @SerialName("tx_pos") val outputIndex: Int, val value: Long, @SerialName("height") val blockHeight: Long) {
    val outPoint by lazy { OutPoint(txid.reversed(), outputIndex.toLong()) }
}
```

Reply:
Looks great!

Reply:
Had to do it differently since we're not actually using kotlin's json serializon. Fixed in 6a66e25.

Reply:
Right, that's even better!